### Install Required Libraries

In [1]:
!pip install chromadb==0.4.0

In [2]:
import chromadb

print(f"Current chromadb version: {chromadb.__version__}")

Current chromadb version: 0.4.0


In [3]:
!pip install -r "C:\Users\nikhi\Downloads\requirements (1).txt"

  Obtaining dependency information for protobuf~=3.19.0 from https://files.pythonhosted.org/packages/32/27/1141a8232723dcb10a595cc0ce4321dcbbd5215300bf4acfc142343205bf/protobuf-3.19.6-py2.py3-none-any.whl.metadata
  Using cached protobuf-3.19.6-py2.py3-none-any.whl.metadata (828 bytes)
Using cached protobuf-3.19.6-py2.py3-none-any.whl (162 kB)
  Attempting uninstall: protobuf
    Found existing installation: protobuf 4.25.3
    Uninstalling protobuf-4.25.3:
      Successfully uninstalled protobuf-4.25.3


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
grpcio-status 1.62.2 requires protobuf>=4.21.6, but you have protobuf 3.19.6 which is incompatible.
tensorflow-intel 2.15.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.19.6 which is incompatible.


In [4]:
!pip install google-generativeai

  Obtaining dependency information for protobuf from https://files.pythonhosted.org/packages/ad/6e/1bed3b7c904cc178cb8ee8dbaf72934964452b3de95b7a63412591edb93c/protobuf-4.25.3-cp310-abi3-win_amd64.whl.metadata
  Using cached protobuf-4.25.3-cp310-abi3-win_amd64.whl.metadata (541 bytes)
Using cached protobuf-4.25.3-cp310-abi3-win_amd64.whl (413 kB)
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.19.6
    Uninstalling protobuf-3.19.6:
      Successfully uninstalled protobuf-3.19.6


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
streamlit 1.22.0 requires protobuf<4,>=3.12, but you have protobuf 4.25.3 which is incompatible.
tensorboard 2.15.1 requires protobuf<4.24,>=3.19.6, but you have protobuf 4.25.3 which is incompatible.


### Google Palm LLM & API key setup

In [5]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyDlXYnP2xNNa0DNa7dPN89u2L4IuAchEg4'

llm = GooglePalm(google_api_key=api_key, temperature=0.2)

### Connect with Database ( MySQL - Workbench)

In [6]:
# Import required libraries
from langchain.utilities import SQLDatabase
from langchain_experimental.sql import SQLDatabaseChain

In [7]:
# Connect with database
db_user = "root"
db_password = "Harish%Nikki98"
db_host = "localhost"
db_name = "car_inv"

db = SQLDatabase.from_uri(f"mysql+pymysql://{db_user}:{db_password}@{db_host}/{db_name}",sample_rows_in_table_info=3)

print(db.table_info)


CREATE TABLE cars (
	car_id INTEGER NOT NULL AUTO_INCREMENT, 
	brand ENUM('Ford','Honda','BMW','Audi') NOT NULL, 
	color ENUM('Red','Blue','Black','White') NOT NULL, 
	model ENUM('SUV','Sedan','Hatchback','Minivan','Sportscar') NOT NULL, 
	price DECIMAL(10, 2), 
	stock_quantity INTEGER NOT NULL, 
	PRIMARY KEY (car_id), 
	CONSTRAINT cars_chk_1 CHECK ((`price` between 10000.00 and 100000.00))
)ENGINE=InnoDB COLLATE utf8mb4_0900_ai_ci DEFAULT CHARSET=utf8mb4

/*
3 rows from cars table:
car_id	brand	color	model	price	stock_quantity
1	Ford	Red	SUV	29778.60	75
2	Audi	Black	SUV	37609.61	57
3	BMW	White	Sportscar	17977.62	10
*/


CREATE TABLE discounts (
	discount_id INTEGER NOT NULL AUTO_INCREMENT, 
	car_id INTEGER NOT NULL, 
	pct_discount DECIMAL(5, 2), 
	PRIMARY KEY (discount_id), 
	CONSTRAINT discounts_ibfk_1 FOREIGN KEY(car_id) REFERENCES cars (car_id), 
	CONSTRAINT discounts_chk_1 CHECK ((`pct_discount` between 0 and 100))
)ENGINE=InnoDB COLLATE utf8mb4_0900_ai_ci DEFAULT CHARSET=utf8mb4

### Ask some questions into database

#### Question 1

In [8]:
cars_db_chain = SQLDatabaseChain.from_llm(llm, db, verbose=True)

# Question in natural language
qns1 = cars_db_chain("How many cars do we have left for ford in sedan model and white color?")



> Entering new SQLDatabaseChain chain...
How many cars do we have left for ford in sedan model and white color?
SQLQuery:SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'
SQLResult: [(74,)]
Answer:74
> Finished chain.


In [9]:
# Sqery qns1 onto the database table
qns1 = cars_db_chain.run("SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'")



> Entering new SQLDatabaseChain chain...
SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'
SQLQuery:SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'
SQLResult: [(74,)]
Answer:74
> Finished chain.


Both the "qns1" responses are same. Model gave the accurate answer for the above question

#### Question 2

In [10]:
qns2 = cars_db_chain.run("How much is the price of the inventory for all SUV model cars?")



> Entering new SQLDatabaseChain chain...
How much is the price of the inventory for all SUV model cars?
SQLQuery:SELECT SUM(price) FROM cars WHERE model = 'SUV'
SQLResult: [(Decimal('538778.15'),)]
Answer:538778.15
> Finished chain.


It made a mistake here. The price is actually the price per unit but in real life database columns will not have perfect names. We need to tell it somehow that price is price per unit and the actual query should be,

SELECT SUM(price*stock_quantity) FROM t_shirts WHERE size = 'S'

In [11]:
qns2 = cars_db_chain.run("SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'")



> Entering new SQLDatabaseChain chain...
SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'
SQLQuery:SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'
SQLResult: [(Decimal('34196152.54'),)]
Answer:34196152.54
> Finished chain.


For question 2, model gave wrong response. Model assumed the price quantity is for the entire stock items.
Modle Generated Query : "SELECT SUM(price) FROM cars WHERE model = 'SUV'"
Actual Query : "SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'"

#### Question 3

In [12]:
qns3 = cars_db_chain.run("If we have to sell all the BMW cars today with discounts applied. How much revenue our store will generate (post discounts)?")



> Entering new SQLDatabaseChain chain...
If we have to sell all the BMW cars today with discounts applied. How much revenue our store will generate (post discounts)?
SQLQuery:SELECT SUM(price - price * pct_discount) FROM cars AS c JOIN discounts AS d ON c.car_id = d.car_id WHERE c.brand = 'BMW' AND d.pct_discount IS NOT NULL AND CURDATE() BETWEEN d.start_date AND d.end_date

OperationalError: (pymysql.err.OperationalError) (1054, "Unknown column 'd.start_date' in 'where clause'")
[SQL: SELECT SUM(price - price * pct_discount) FROM cars AS c JOIN discounts AS d ON c.car_id = d.car_id WHERE c.brand = 'BMW' AND d.pct_discount IS NOT NULL AND CURDATE() BETWEEN d.start_date AND d.end_date]
(Background on this error at: https://sqlalche.me/e/14/e3q8)

Above, it returned a wrong query which generated an error during query execution. It thinks discount
table would have start and end date which is normally true but in our table there is no start or end date column.
One thing we can do here is run the query directly.

In [13]:
sql_code = """
select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'BMW'
group by car_id) a left join discounts on a.car_id = discounts.car_id
 """

qns3 = cars_db_chain.run(sql_code)



> Entering new SQLDatabaseChain chain...

select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'BMW'
group by car_id) a left join discounts on a.car_id = discounts.car_id
 
SQLQuery:select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'BMW'
group by car_id) a left join discounts on a.car_id = discounts.car_id
SQLResult: [(Decimal('61370833.35000000'),)]
Answer:61370833.35
> Finished chain.


Above, for question 3 model returned a wrong query which generated an error during query execution. It thinks discount
table would have start and end date which is normally true but in the car table there is no start or end date column.

It produced a correct answer when explicit query was given. 61370833.35 is the correct response.

Later will use this same query on for few shot learning, and compare the result.

#### Question 4

In [14]:
qns4 = cars_db_chain.run("How much revenue  our store will generate by selling all Ford cars without discount?")



> Entering new SQLDatabaseChain chain...
How much revenue  our store will generate by selling all Ford cars without discount?
SQLQuery:SELECT SUM(price) FROM cars WHERE brand = 'Ford' AND car_id NOT IN (SELECT car_id FROM discounts)
SQLResult: [(Decimal('49731.79'),)]
Answer:49731.79
> Finished chain.


In [15]:
qns4 = cars_db_chain.run("SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford'")



> Entering new SQLDatabaseChain chain...
SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford'
SQLQuery:SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford'
SQLResult: [(Decimal('42061677.20'),)]
Answer:42061677.20
> Finished chain.


Model generated qrong answer for the above query. Same as question 2, model assumes the price is for the entire stock items and also it applies discount of the items where the question asked is not to apply discount.
Modle generated Query : "SELECT SUM(price) FROM cars WHERE brand = 'Ford' AND car_id NOT IN (SELECT car_id FROM discounts)"
Actual Quesry : "SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford'"

#### Question 5

In [16]:
qns5 = cars_db_chain.run("How many red color audi cars we have available?")



> Entering new SQLDatabaseChain chain...
How many red color audi cars we have available?
SQLQuery:SELECT count(*) FROM cars WHERE color = 'Red' AND brand = 'Audi'
SQLResult: [(4,)]
Answer:4
> Finished chain.


Once again above is the wrong answer. We need to use SUM(stock_quantity). Let's run the query explicitly. We will use the result of this query later on in the notebook

In [17]:
qns5 = cars_db_chain.run("SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'")



> Entering new SQLDatabaseChain chain...
SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'
SQLQuery:SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'
SQLResult: [(Decimal('242'),)]
Answer:242
> Finished chain.


Model generated wrong response again. 

Model generated query : "SELECT count(*) FROM cars WHERE color = 'Red' AND brand = 'Audi'"
Actual Query : "SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'"

## Few Shot Learning

##### We will use few shot learning to fix issues we have seen so far.


##### Few-shot learning is a machine learning approach where a model is trained to generalize from a very limited amount of labeled data

In [18]:
# Creating few_shorts labeled training data to train model for our specific use case
few_shots = [
    {'Question': "How many cars do we have left for ford in sedan model and white color?",
     'SQLQuery': "SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'",
     'SQLResult': "Result of the SQL query",
     'Answer': "qns1"},
    {'Question': "How much is the price of the inventory for all SUV model cars?",
     'SQLQuery': "SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'",
     'SQLResult': "Result of the SQL query",
     'Answer': "qns2"},
    {'Question': "If we have to sell all the BMW cars today with discounts applied. How much revenue our store will generate (post discounts)?",
     'SQLQuery': """select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
					(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'BMW'
					group by car_id) a left join discounts on a.car_id = discounts.car_id""",
     'SQLResult': "Result of the SQL query",
     'Answer': "qns3"},
    {'Question': "How much revenue  our store will generate by selling all Ford cars without discount?",
     'SQLQuery': "SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford'",
     'SQLResult': "Result of the SQL query",
     'Answer': "qns4"},
    {'Question': "How many red color audi cars we have available?",
     'SQLQuery': "SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'",
     'SQLResult': "Result of the SQL query",
     'Answer': "qns5"}
]


## Creating Semantic Similarity Based example selector

#### create embedding on the few_shots
#### Store the embeddings in Chroma DB
#### Retrieve the the top most Semantically close example from the vector store

In [19]:
# Import necessary libraries
from langchain.prompts import SemanticSimilarityExampleSelector
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Chroma

# Create embeddings
embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2')

# Convert the few_shorts into list of strings
to_vectorize = [" ".join(example.values()) for example in few_shots]

C:\Users\nikhi\anaconda3\Lib\site-packages\transformers\utils\generic.py:260: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  torch.utils._pytree._register_pytree_node(


In [20]:
to_vectorize

["How many cars do we have left for ford in sedan model and white color? SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White' Result of the SQL query qns1",
 "How much is the price of the inventory for all SUV model cars? SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV' Result of the SQL query qns2",
 "If we have to sell all the BMW cars today with discounts applied. How much revenue our store will generate (post discounts)? select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from\n\t\t\t\t\t(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'BMW'\n\t\t\t\t\tgroup by car_id) a left join discounts on a.car_id = discounts.car_id Result of the SQL query qns3",
 "How much revenue  our store will generate by selling all Ford cars without discount? SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Ford' Result of the SQL query qns4",
 "How many red color aud

## Store Embeddings into Chroma DB

In [21]:
# Store embeddings into chroma db
vectorstore = Chroma.from_texts(to_vectorize, embeddings, metadatas=few_shots)

In [22]:
exp_selector = SemanticSimilarityExampleSelector(
    vectorstore=vectorstore,
    k=2,
)

exp_selector.select_examples({"Question": "How many Audi cars have left in my store?"})

[{'Answer': 'qns5',
  'Question': 'How many red color audi cars we have available?',
  'SQLQuery': "SELECT sum(stock_quantity) FROM cars WHERE brand = 'Audi' AND color = 'Red'",
  'SQLResult': 'Result of the SQL query'},
 {'Answer': 'qns1',
  'Question': 'How many cars do we have left for ford in sedan model and white color?',
  'SQLQuery': "SELECT stock_quantity FROM cars WHERE brand = 'Ford' AND model = 'Sedan' AND color = 'White'",
  'SQLResult': 'Result of the SQL query'}]

In [23]:
### my sql based instruction prompt
mysql_prompt = """You are a MySQL expert. Given an input question, first create a syntactically correct MySQL query to run, then look at the results of the query and return the answer to the input question.
Unless the user specifies in the question a specific number of examples to obtain, query for at most {top_k} results using the LIMIT clause as per MySQL. You can order the results to return the most informative data in the database.
Never query for all columns from a table. You must query only the columns that are needed to answer the question. Wrap each column name in backticks (`) to denote them as delimited identifiers.
Pay attention to use only the column names you can see in the tables below. Be careful to not query for columns that do not exist. Also, pay attention to which column is in which table.
Pay attention to use CURDATE() function to get the current date, if the question involves "today".

Use the following format:

Question: Question here
SQLQuery: Query to run with no pre-amble
SQLResult: Result of the SQLQuery
Answer: Final answer here

No pre-amble.
"""

In [24]:
from langchain.prompts import FewShotPromptTemplate
from langchain.chains.sql_database.prompt import PROMPT_SUFFIX, _mysql_prompt

print(PROMPT_SUFFIX)

Only use the following tables:
{table_info}

Question: {input}


## Setting up PromptTemplete using input variables

In [25]:
from langchain.prompts.prompt import PromptTemplate

example_prompt = PromptTemplate(
    input_variables=["Question", "SQLQuery", "SQLResult","Answer",],
    template="\nQuestion: {Question}\nSQLQuery: {SQLQuery}\nSQLResult: {SQLResult}\nAnswer: {Answer}",
)

In [26]:
print(_mysql_prompt)

You are a MySQL expert. Given an input question, first create a syntactically correct MySQL query to run, then look at the results of the query and return the answer to the input question.
Unless the user specifies in the question a specific number of examples to obtain, query for at most {top_k} results using the LIMIT clause as per MySQL. You can order the results to return the most informative data in the database.
Never query for all columns from a table. You must query only the columns that are needed to answer the question. Wrap each column name in backticks (`) to denote them as delimited identifiers.
Pay attention to use only the column names you can see in the tables below. Be careful to not query for columns that do not exist. Also, pay attention to which column is in which table.
Pay attention to use CURDATE() function to get the current date, if the question involves "today".

Use the following format:

Question: Question here
SQLQuery: SQL Query to run
SQLResult: Result of

In [40]:
few_shot_prompt = FewShotPromptTemplate(
    example_selector=exp_selector,
    example_prompt=example_prompt,
    prefix=mysql_prompt,
    suffix=PROMPT_SUFFIX,
    input_variables=["input", "table_info", "top_k"], 
)

In [41]:
new_db_chain = SQLDatabaseChain.from_llm(llm, db, verbose=True, prompt=few_shot_prompt)

Now this is working ok. Previously for this same question it was giving wrong answer because it did not use SUM clause around stock_quantity column

In [42]:
new_db_chain("How much is the price of the inventory for all SUV model cars")



> Entering new SQLDatabaseChain chain...
How much is the price of the inventory for all SUV model cars
SQLQuery:SELECT SUM(price*stock_quantity) FROM cars WHERE model = 'SUV'
SQLResult: [(Decimal('34196152.54'),)]
Answer:34196152.54
> Finished chain.


{'query': 'How much is the price of the inventory for all SUV model cars',
 'result': '34196152.54'}

Now this is working ok. Previously for this same question it was giving wrong answer because it did not use SUM clause around stock_quantity column

In [43]:
new_db_chain("How much is the price of all black color Honda cars?")



> Entering new SQLDatabaseChain chain...
How much is the price of all black color Honda cars?
SQLQuery:SELECT sum(price) FROM cars WHERE brand = 'Honda' AND color = 'Black'
SQLResult: [(Decimal('143469.36'),)]
Answer:143469.36
> Finished chain.


{'query': 'How much is the price of all black color Honda cars?',
 'result': '143469.36'}

#### The model is now summing up all the result price column

In [44]:
new_db_chain("If we have to sell all the Ford cars today with discounts applied. How much revenue our store will generate (post discounts)?")



> Entering new SQLDatabaseChain chain...
If we have to sell all the Ford cars today with discounts applied. How much revenue our store will generate (post discounts)?
SQLQuery:select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
					(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'Ford'
					group by car_id) a left join discounts on a.car_id = discounts.car_id
SQLResult: [(Decimal('19706712.96973900'),)]
Answer:19706712.969739
> Finished chain.


{'query': 'If we have to sell all the Ford cars today with discounts applied. How much revenue our store will generate (post discounts)?',
 'result': '19706712.969739'}

In [35]:
#new_db_chain("If we have to sell all the Audi cars today with discounts applied. How much revenue  our store will generate (post discounts)?")



> Entering new SQLDatabaseChain chain...
If we have to sell all the Audi cars today with discounts applied. How much revenue  our store will generate (post discounts)?
SQLQuery:select sum(a.total_amount * ((100-COALESCE(discounts.pct_discount,0))/100)) as total_revenue from
					(select sum(price*stock_quantity) as total_amount, car_id from cars where brand = 'Audi'
					group by car_id) a left join discounts on a.car_id = discounts.car_id
SQLResult: [(Decimal('51912355.50000000'),)]
Answer:qns1
> Finished chain.


{'query': 'If we have to sell all the Audi cars today with discounts applied. How much revenue  our store will generate (post discounts)?',
 'result': 'qns1'}

In [45]:
new_db_chain.run('How much revenue  our store will generate by selling all Audi cars of hatch back model without discount?')



> Entering new SQLDatabaseChain chain...
How much revenue  our store will generate by selling all Audi cars of hatch back model without discount?
SQLQuery:SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Audi' AND model = 'Hatchback'
SQLResult: [(Decimal('5389999.33'),)]
Answer:qns1
> Finished chain.


'qns1'

In [46]:
new_db_chain.run('How much revenue can i generate by selling all honda black cars with minivan model?')



> Entering new SQLDatabaseChain chain...
How much revenue can i generate by selling all honda black cars with minivan model?
SQLQuery:SELECT SUM(price * stock_quantity) FROM cars WHERE brand = 'Honda' AND color = 'Black' AND model = 'Minivan'
SQLResult: [(Decimal('1231515.25'),)]
Answer:qns1
> Finished chain.


'qns1'